File: android/app/src/main/java/org/team8bitpool/app/core/LessonMatch.kt
package org.team8bitpool.app.core

/**
 * Port of lesson_match.py: match a recognised utterance to the content pack's
 * pre-translated lines (cosine similarity of character trigram counts of the
 * normalised text, numbers written as words; a match also needs the line's exact
 * numbers). Tested against
 * tests/data/lesson_match_vectors.json (LessonMatchTest). Thresholds:
 * config.LESSON_MATCH_THRESHOLD, tuned in bench/results/lesson_match.md.
 */
object LessonMatch {
    val HINDI_0_99: List<String> = ("शून्य एक दो तीन चार पांच छह सात आठ नौ दस ग्यारह बारह तेरह चौदह पंद्रह सोलह सत्रह अठारह उन्नीस " +
        "बीस इक्कीस बाईस तेईस चौबीस पच्चीस छब्बीस सत्ताईस अट्ठाईस उनतीस तीस इकतीस बत्तीस तैंतीस चौंतीस " +
        "पैंतीस छत्तीस सैंतीस अड़तीस उनतालीस चालीस इकतालीस बयालीस तैंतालीस चवालीस पैंतालीस छियालीस सैंतालीस " +
        "अड़तालीस उनचास पचास इक्यावन बावन तिरेपन चौवन पचपन छप्पन सत्तावन अट्ठावन उनसठ साठ इकसठ बासठ " +
        "तिरेसठ चौंसठ पैंसठ छियासठ सड़सठ अड़सठ उनहत्तर सत्तर इकहत्तर बहत्तर तिहत्तर चौहत्तर पचहत्तर " +
        "छिहत्तर सतहत्तर अठहत्तर उन्यासी अस्सी इक्यासी बयासी तिरासी चौरासी पचासी छियासी सत्तासी अट्ठासी " +
        "नवासी नब्बे इक्यानवे बानवे तिरानवे चौरानवे पचानवे छियानवे सत्तानवे अट्ठानवे निन्यानवे").split(" ")

    /** Digit by digit, for numbers too long for the word forms (as Python does). */
    private fun digitWise(digits: String, lang: String) =
        digits.map { if (lang == "hi") HINDI_0_99[it - '0'] else OlChiki.NUMBER_WORDS.getValue(it - '0') }.joinToString(" ")

    fun hindiNumber(digits: String): String {
        val s = digits.trimStart('0').ifEmpty { "0" }
        if (s.length > 4) return digitWise(s, "hi")
        val n = s.toInt()
        if (n < 100) return HINDI_0_99[n]
        val th = n / 1000; val hu = (n % 1000) / 100; val r = n % 100
        val parts = ArrayList<String>()
        if (th != 0) { parts += HINDI_0_99[th]; parts += "हजार" }
        if (hu != 0) { parts += HINDI_0_99[hu]; parts += "सौ" }
        if (r != 0) parts += HINDI_0_99[r]
        return parts.joinToString(" ")
    }

    private fun numberWords(digits: String, lang: String): String {
        if (lang == "hi") return hindiNumber(digits)
        val s = digits.trimStart('0').ifEmpty { "0" }
        return if (s.length > 3) digitWise(s, "sat") else OlChiki.numberToSantali(s.toLong())
    }

    fun canonical(text: String, lang: String): String {
        val k = TextNorm.key(text)
        return TextNorm.key(Regex("[0-9]+").replace(k) { " ${numberWords(it.value, lang)} " })
    }

    private val HI_NUMBER_WORDS: Set<String> by lazy { (HINDI_0_99 + listOf("सौ", "हजार")).map { TextNorm.key(it) }.toSet() }
    private val SAT_NUMBER_WORDS: Set<String> by lazy { OlChiki.NUMBER_WORDS.values.map { TextNorm.key(it) }.toSet() }

    /** The number words of canonical(text), in order (port of lesson_match.number_tokens). */
    fun numberTokens(text: String, lang: String): List<String> {
        val words = if (lang == "hi") HI_NUMBER_WORDS else SAT_NUMBER_WORDS
        return canonical(text, lang).split(" ").filter { it in words }
    }

    fun trigrams(text: String, lang: String): Map<String, Int> {
        val s = " ${canonical(text, lang)} "
        val m = HashMap<String, Int>()
        // code points, as Python slices strings (every script used here is in the BMP)
        for (i in 0..s.length - 3) m.merge(s.substring(i, i + 3), 1, Int::plus)
        return m
    }

    private fun norm(t: Map<String, Int>) = Math.sqrt(t.values.sumOf { it.toDouble() * it })

    fun similarity(a: String, b: String, lang: String): Double {
        val ta = trigrams(a, lang); val tb = trigrams(b, lang)
        if (ta.isEmpty() || tb.isEmpty()) return 0.0
        val dot = ta.entries.sumOf { (k, v) -> v.toDouble() * (tb[k] ?: 0) }
        return dot / Math.sqrt(norm(ta) * norm(ta) * norm(tb) * norm(tb))
    }

    data class Result(val line: String?, val score: Double, val matched: Boolean, val numbersAgree: Boolean = false)

    /** The closest candidate (ties: the first in order), matched only when score >= threshold and the numbers agree. */
    fun match(text: String, candidates: Iterable<String>, threshold: Double, lang: String): Result {
        val t = trigrams(text, lang)
        if (t.isEmpty()) return Result(null, 0.0, false)
        val nt = norm(t)
        var top: String? = null; var score = 0.0
        for (c in candidates) {
            val tc = trigrams(c, lang)
            if (tc.isEmpty()) continue
            val s = t.entries.sumOf { (k, v) -> v.toDouble() * (tc[k] ?: 0) } / (nt * norm(tc))
            if (s > score) { top = c; score = s }
        }
        val agree = top != null && numberTokens(text, lang) == numberTokens(top, lang)
        return Result(top, score, top != null && score >= threshold && agree, agree)
    }
}

File: lesson_match.py
"""Curriculum matching for the tablet's on-device voice path (A1).

The tablet has speech recognition but (until M4) no translation model. A spoken
lesson line is recognised, then matched against the content pack's pre-translated
lines; the pack's translation and audio are used only when the match is close
enough. Below the threshold the page shows the transcript and "not a lesson line"
and never invents a translation.

Similarity: cosine similarity of character trigram counts of textnorm.normalize_key
(text), padded with one space at each end, so word boundaries count. 1.0 = the same
trigrams in the same proportions. Numbers are compared as words: a digit run is
written out (Hindi words for Hindi, translit.olchiki.number_to_santali for Santali)
before the trigrams are taken, because a line written "4" and a line written "चार"
are the same utterance and speech recognition writes either. A match also needs the
same numbers in the same order as the line: in a maths lesson "दो" for "पांच" is a
different sentence, however similar the rest. The Kotlin port (android .../core/LessonMatch.kt)
is tested against tests/data/lesson_match_vectors.json, written from this file.

The threshold (config.LESSON_MATCH_THRESHOLD) is tuned by bench/lesson_match_tune.py
on synthetic lesson clips and scored on a held-out half: bench/results/lesson_match.md.
"""

import math
import re
from collections import Counter

from textnorm import normalize_key

# 0-99 in Hindi (standard spellings; normalize_key folds nukta and chandrabindu).
HINDI_0_99 = ("शून्य एक दो तीन चार पांच छह सात आठ नौ दस ग्यारह बारह तेरह चौदह पंद्रह सोलह सत्रह अठारह उन्नीस "
              "बीस इक्कीस बाईस तेईस चौबीस पच्चीस छब्बीस सत्ताईस अट्ठाईस उनतीस तीस इकतीस बत्तीस तैंतीस चौंतीस "
              "पैंतीस छत्तीस सैंतीस अड़तीस उनतालीस चालीस इकतालीस बयालीस तैंतालीस चवालीस पैंतालीस छियालीस सैंतालीस "
              "अड़तालीस उनचास पचास इक्यावन बावन तिरेपन चौवन पचपन छप्पन सत्तावन अट्ठावन उनसठ साठ इकसठ बासठ "
              "तिरेसठ चौंसठ पैंसठ छियासठ सड़सठ अड़सठ उनहत्तर सत्तर इकहत्तर बहत्तर तिहत्तर चौहत्तर पचहत्तर "
              "छिहत्तर सतहत्तर अठहत्तर उन्यासी अस्सी इक्यासी बयासी तिरासी चौरासी पचासी छियासी सत्तासी अट्ठासी "
              "नवासी नब्बे इक्यानवे बानवे तिरानवे चौरानवे पचानवे छियानवे सत्तानवे अट्ठानवे निन्यानवे").split()
assert len(HINDI_0_99) == 100
_HI_NUMBER_WORDS = {normalize_key(w) for w in HINDI_0_99} | {"सौ", "हजार"}


def hindi_number(n: int) -> str:
    """0-9999 in Hindi words; larger numbers digit by digit."""
    if n < 100:
        return HINDI_0_99[n]
    if n >= 10000:
        return " ".join(HINDI_0_99[int(d)] for d in str(n))
    th, r = divmod(n, 1000)
    hu, r = divmod(r, 100)
    parts = ([HINDI_0_99[th], "हजार"] if th else []) + ([HINDI_0_99[hu], "सौ"] if hu else [])
    return " ".join(parts + ([HINDI_0_99[r]] if r else []))


def number_words(n: int, lang: str) -> str:
    if lang == "hi":
        return hindi_number(n)
    from translit.olchiki import number_to_santali
    return number_to_santali(n)


def canonical(text: str, lang: str) -> str:
    """normalize_key with every digit run written out in words."""
    k = normalize_key(text)
    return normalize_key(re.sub(r"[0-9]+", lambda m: f" {number_words(int(m.group()), lang)} ", k))


_SAT_NUMBER_WORDS = None


def number_tokens(text: str, lang: str) -> list:
    """The number words in canonical(text), in order (digits already written as words)."""
    global _SAT_NUMBER_WORDS
    if lang == "hi":
        words = _HI_NUMBER_WORDS
    else:
        if _SAT_NUMBER_WORDS is None:
            from translit.olchiki import _number_words
            _SAT_NUMBER_WORDS = {normalize_key(w) for w in _number_words().values()}
        words = _SAT_NUMBER_WORDS
    return [w for w in canonical(text, lang).split() if w in words]


def trigrams(text: str, lang: str = "hi") -> Counter:
    s = f" {canonical(text, lang)} "
    return Counter(s[i:i + 3] for i in range(len(s) - 2)) if len(s) >= 3 else Counter()


def similarity(a: str, b: str, lang: str = "hi") -> float:
    ta, tb = trigrams(a, lang), trigrams(b, lang)
    if not ta or not tb:
        return 0.0
    dot = sum(v * tb.get(k, 0) for k, v in ta.items())
    return dot / math.sqrt(sum(v * v for v in ta.values()) * sum(v * v for v in tb.values()))


def best(text: str, candidates, lang: str = "hi"):
    """(candidate, score) of the closest candidate, or (None, 0.0). Ties: the first in order."""
    top, score = None, 0.0
    t = trigrams(text, lang)
    if not t:
        return None, 0.0
    nt = math.sqrt(sum(v * v for v in t.values()))
    for c in candidates:
        tc = trigrams(c, lang)
        if not tc:
            continue
        s = sum(v * tc.get(k, 0) for k, v in t.items()) / (nt * math.sqrt(sum(v * v for v in tc.values())))
        if s > score:
            top, score = c, s
    return top, score


def match(text: str, candidates, threshold: float, lang: str = "hi"):
    """{"line", "score", "matched", "numbers_agree"}: matched only when score >= threshold
    and the utterance has exactly the line's numbers (so "दो आम" never matches "पांच आम")."""
    line, score = best(text, candidates, lang)
    agree = line is not None and number_tokens(text, lang) == number_tokens(line, lang)
    return {"line": line, "score": round(score, 6), "numbers_agree": agree,
            "matched": line is not None and score >= threshold and agree}
